nav2_costmap_2d/plugins/costmap_filters/costmap_filter.cpp

```diff
@@ -0,0 +1,119 @@
+/*********************************************************************
+ *
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2020 Samsung Research Russia
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the <ORGANIZATION> nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Alexey Merzlyakov
+ *********************************************************************/
+
+#include "nav2_costmap_2d/costmap_filters/costmap_filter.hpp"
+
+#include <exception>
+
+namespace nav2_costmap_2d
+{
+
+CostmapFilter::CostmapFilter()
+: is_rolling_(true), costmap_filter_info_topic_("/costmap_filter_info")
+{
+}
+
+CostmapFilter::~CostmapFilter()
+{
+}
+
+void CostmapFilter::onInitialize()
+{
+  // Get parameters
+  declareParameter("enabled", rclcpp::ParameterValue(true));
```

Comment:
Move declares in the constructor. If we initialize multiple times this will crash

Reply:
If twice call `declareParameter()`, it will add a parameter in the `local_params_` only once, because this is a `std::unordered_set`. Also, it calls `nav2_util::declare_parameter_if_not_declared()` routine which has a protection from second call. So, this should be safe. Also, all other plugins are containing `declareParameter()` inside `onInitialize()`. Do I really need to move it into constructor?

Reply:
Got it